AppTests/App/Logic/DataUploadLogicTests.swift

```diff
@@ -313,26 +318,498 @@ final class DataUploadLogicTests: XCTestCase {
+    let dependencies = AppDependencies.mocked(
+      getAppState: { state },
+      dispatch: dispatchInterceptor.dispatchFunction,
+      exponentialDistributionGenerator: DeterministicGenerator.self
+    )
+
+    let context = AppSideEffectContext(dependencies: dependencies)
+
+    try Logic.DataUpload.ScheduleDummyIngestionSequenceIfNecessary().sideEffect(context)
+
+    try XCTAssertNotContainsType(dispatchInterceptor.dispatchedItems, DelayedDispatchable.self)
+  }
+}
+
+// MARK: - Dummy traffic scheduling
+
+extension DataUploadLogicTests {
+  func testSimulationDoesNotStartIfSessionCancelled() throws {
+    var state = AppState()
+    state.ingestion.isDummyTrafficSequenceCancelled = true
+
+    let requestExecutor = MockRequestExecutor(mockedResult: .success(Data()))
+
+    let dispatchInterceptor = DispatchInterceptor()
+    let dependencies = AppDependencies.mocked(
+      getAppState: { state },
+      dispatch: dispatchInterceptor.dispatchFunction,
+      requestExecutor: requestExecutor,
+      exponentialDistributionGenerator: DeterministicGenerator.self
+    )
+
+    let context = AppSideEffectContext(dependencies: dependencies)
+
+    try Logic.DataUpload.StartIngestionSequenceIfNotCancelled().sideEffect(context)
+
+    XCTAssertEqual(requestExecutor.executeMethodCalls.count, 0)
+  }
+
+  func testSimulationUpdatesOpportunityWindowIfCancelled() throws {
+    var state = AppState()
+    state.ingestion.isDummyTrafficSequenceCancelled = true
+
+    let dispatchInterceptor = DispatchInterceptor()
+    let dependencies = AppDependencies.mocked(
+      getAppState: { state },
+      dispatch: dispatchInterceptor.dispatchFunction
+    )
+
+    let context = AppSideEffectContext(dependencies: dependencies)
+
+    try Logic.DataUpload.StartIngestionSequenceIfNotCancelled().sideEffect(context)
+
+    try XCTAssertContainsType(dispatchInterceptor.dispatchedItems, Logic.DataUpload.UpdateDummyTrafficOpportunityWindow.self)
+  }
+
+  func testSimulationSendsOneRequestEvenIfProbabilityIsZero() throws {
+    var state = AppState()
+    state.ingestion.isDummyTrafficSequenceCancelled = false
+
+    let requestExecutor = MockRequestExecutor(mockedResult: .success(Data()))
+
+    DeterministicGenerator.randomValue = 1
+
+    let dispatchInterceptor = DispatchInterceptor()
+    let dependencies = AppDependencies.mocked(
+      getAppState: { state },
+      dispatch: dispatchInterceptor.dispatchFunction,
+      requestExecutor: requestExecutor,
+      uniformDistributionGenerator: DeterministicGenerator.self
+    )
+
+    let context = AppSideEffectContext(dependencies: dependencies)
+
+    try Logic.DataUpload.StartIngestionSequenceIfNotCancelled().sideEffect(context)
+
+    XCTAssertEqual(requestExecutor.executeMethodCalls.count, 1)
+  }
+
+  func testSimulationUpdatesOpportunityWindowIfNotCancelled() throws {
+    var state = AppState()
+    state.ingestion.isDummyTrafficSequenceCancelled = false
+
+    /// Fail the first dice roll
```

Comment:
<!--
  0 failure: 
  1 warning:  swiftlint (0.39.2...
  
  
  DangerID: danger-id-Danger;
  File: AppTests/App/Logic/DataUploadLogicTests.swift;
  Line: 744;
-->

- :warning: swiftlint (0.39.2): A doc comment should be attached to a declaration. (orphaned_doc_comment)